### PR TITLE
Add line geometry

### PR DESCRIPTION
**Pull request by bcvery1**
This adds a new geometric shape, a `Line`, made of two `Vec`s

Methods include intersections with other `Line`s, `Rect`s and `Circle`s (as well as visa-versa), moving, rotating and others.  All functions have tests for them.

There was discussion about whether you wanted a separate repo for collisions, and it's your repo so no problem if that's the decision; personal view: it makes sense to me for this code to be in Pixel.

Please let me know if you'd like anything changed/fixed (although should all be working!) or discussed.

Cheers

### CODE REVIEW

**Comment by bcvery1** [comment]
@faiface Thanks for the positive feedback.  In regards to returning the point of intersection, that sounds useful, but consider this: a slice of Lines make a polygon, when intersect detection is run on a polygon you'd want the same pattern (Vector to prevent the overlap) as Rect and Circle.

Can I suggest as an alternative to your suggestions, that I add a second function (one for each rectangles and circles) which provides the point of intersections (as a line could intersect a shape in one *or* two places).

In regard to your other requested changes, all valid, and I'll get them updated.

**Comment by faiface** [comment]
You definitely have a point.

The idea with another method is good, let's go with that. Just that the name `Intersect` would be best for it :D. If we were to do it all over, this would be the names:

- `FixOverlap` - for a vector to fix overlap.
- `Intersect` - for an intersection point / list of points.

Since Pixel still isn't 1.0 and we use go mod, we are allowed to make breaking changes, so this is a possibility.

Another possibility is to name it differently, for example: `Intersections` or `PointsOfIntersection`.

**Comment by bcvery1** [comment]
I'd suggest making those name changes are out of scope for this change, but also that I think greater consistency with the geometry structs would be beneficial.  Not least in that some meaningful interfaces could be created where you could have a collection of objects, and check if each of them collide with a `Rect` without caring about their specific types...

Maybe in a different issue.

I'll probably go with `func (T Type) IntersectPoints(l Line) []Vec`, if you're happy with that?  It's succinct, idiomatic and self-explaining.

**Comment by faiface** [comment]
The renames would be consistent among all geometry shapes in Pixel.

But I'm fine with your proposition, except that let's call it `IntersectionPoints` because that's proper English :D. `IntersectPoints` would mean we are intersecting points.

**Comment by bcvery1** [comment]
Changes pushed for rectangle intersection points (and a few fixes to line-line intersection I found on the way) - circle intersection points will probably be done tomorrow

Not yet ready for merge

File: geometry.go [hunk 1]
```diff
@@ -181,6 +181,278 @@ func Lerp(a, b Vec, t float64) Vec {
...
+
+	if cirToClosest.Len() >= c.Radius {
+		return ZV
+	}
+
+	return cirToClosest.Scaled(cirToClosest.Len() - c.Radius)
+}
+
+// IntersectRect will return the shortest `Vec` such that moving the Line by that Vec will cause  the Line and Rect to
+// no longer intesect.  If they do not intersect at all, this function will return a zero-vector.
```
[the remark this answers is not in the document]
Comment (bcvery1) [reply geometry.go]: Resolved by introducing `func (r Rect) IntersectionPoints(l Line) []Vec`

File: geometry.go [hunk 2]
```diff
@@ -181,6 +181,278 @@ func Lerp(a, b Vec, t float64) Vec {
...
+	if l.Contains(V(x, y)) && k.Contains(V(x, y)) {
+		// The intersect point is on both line segments, they intersect.
+		return V(x, y), true
+	}
+
+	return ZV, false
+}
+
+// IntersectCircle will return the shortest `Vec` such that moving the Line by that Vec will cause the Line and Circle
+// to no longer intesect.  If they do not intersect at all, this function will return a zero-vector.
```
[the remark this answers is not in the document]
Comment (bcvery1) [reply geometry.go]: Resolved by introducing `func (c Circle) IntersectionPoints(l Line) []Vec`

**Comment by bcvery1** [comment]
@faiface All done, I think

File: geometry.go [hunk 3]
```diff
@@ -181,6 +181,268 @@ func Lerp(a, b Vec, t float64) Vec {
 	return a.Scaled(1 - t).Add(b.Scaled(t))
 }
 
+// Line is a 2D line segment, between points `A` and `B`.
```
Comment (faiface) [inline geometry.go]: Forgot backticks

File: geometry.go [hunk 4]
```diff
@@ -181,6 +181,268 @@ func Lerp(a, b Vec, t float64) Vec {
...
+		if toStart.Len() < toEnd.Len() {
+			return l.A
+		}
+		return l.B
+	}
+
+	return V(x, y)
+}
+
+// Contains returns whether the provided Vec lies on the line
```
Comment (faiface) [inline geometry.go]: Forgot period at the end

**Comment by bcvery1** [comment]
Looks like intersection points being returned aren't always in the same order (there is no meaning to the order, so they needn't be) - will put a fix in so it checks the result is in the list, and doesn't check the order

**Comment by faiface** [comment]
Order may actually be very useful. I would order them from the closest to the A point to the furthest and document this behavior. This will be very useful when finding the first intersection of a line with an object.

**Comment by bcvery1** [comment]
Done!

**Comment by faiface** [comment]
The slices can be at most length 2, right? It would be much more performant to check the size and if it's 2 and they are in the wrong order, swap them.

After that, I'll merge, I promise :D

**Comment by bcvery1** [comment]
Is that not what I'm doing?  In `func (r Rect) IntersectionPoints`  I have `if len(points) == 2 {` then check-swap.

In `func (c Circle) IntersectionPoints` it is only check-swapped when there are going to be two points.

**Comment by faiface** [comment]
Oh, it is... I must've been looking at an older commit or something...

Okay, I'm very satisfied with this! Great job, as usual!

**Pull request merged**